compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/KotlinPoet.kt

```diff
@@ -0,0 +1,230 @@
+  callableDeclaration: T
+): TypeName {
+  // If the parameter is annotated with @JvmSuppressWildcards, then add the annotation
+  // to our type so that this information is forwarded when our Factory is compiled.
+  val hasJvmSuppressWildcards =
+    callableDeclaration.typeReference?.hasAnnotation(jvmSuppressWildcardsFqName) ?: false
+
+  // Add the @JvmSuppressWildcards annotation even for simple generic return types like
+  // Set<String>. This avoids some edge cases where Dagger chokes.
+  val isGenericType = callableDeclaration.typeReference?.isGenericType() ?: false
+
+  return if (hasJvmSuppressWildcards || isGenericType) this.jvmSuppressWildcards() else this
+}
+
+internal fun List<Parameter>.asArgumentList(
+  asProvider: Boolean,
+  includeModule: Boolean
+): String {
+  return this
+      .let { list ->
+        if (asProvider) {
+          list.map { parameter ->
+            when {
+              parameter.isWrappedInProvider -> parameter.name
+              parameter.isWrappedInLazy ->
+                "$daggerDoubleCheckFqNameString.lazy(${parameter.name})"
+              else -> "${parameter.name}.get()"
+            }
+          }
+        } else list.map { it.name }
+      }
+      .let {
+        if (includeModule) {
+          val result = it.toMutableList()
+          result.add(0, "module")
+          result.toList()
+        } else {
+          it
+        }
+      }
+      .joinToString()
+}
+
+private fun TypeName.wrapInProvider(): ParameterizedTypeName {
+  return Provider::class.asClassName().parameterizedBy(this)
+}
+
+private fun TypeName.wrapInLazy(): ParameterizedTypeName {
+  return Lazy::class.asClassName().parameterizedBy(this)
+}
+
+internal fun TypeSpec.Builder.addAnvilAnnotation(): TypeSpec.Builder {
+  return addAnnotation(
+      AnnotationSpec.builder(Generated::class)
+          .addMember("value = [%S]", AnvilComponentRegistrar::class.java.canonicalName)
+          .addMember("comments = %S", "https://github.com/square/anvil")
+          .build()
+  )
+}
+
+internal fun FileSpec.writeToString(): String {
+  val stream = ByteArrayOutputStream()
+  stream.writer().use {
+    writeTo(it)
+  }
+  return stream.toString()
+}
+
+/**
+ * Removes invalid imports that can be caused by [requireTypeName] and copies imports from the
+ * source file. Copying imports is necessary for star imports.
+ */
+internal fun String.replaceImports(clazz: KtClassOrObject): String {
+  val originalLines = lines()
+  val linesWithoutImports = originalLines
+      .filterNot { it.startsWith("import ") }
+      .toMutableList()
+
+  val copiedImports = clazz.containingKtFile
+      .importDirectives
+      .map { it.text.trim() }
+
+  val importDirectives = copiedImports
+      .plus(
+          originalLines
+              .filter {
+                it.startsWith("import ") && it.contains(".")
+              }
+              .map { it.trim() }
+      )
+      .distinct()
+      .toMutableList()
```

Comment:
Do we care about order at all here?

Reply:
The order is important for the build cache. In theory it shouldn't matter, because it doesn't impact the ABI, but with the Kotlin plugin you never know...

I think the order is stable. But to be sure the imports should be sorted before they're written to the file.